discord/ext/commands/converter.py

```diff
@@ -159,6 +159,41 @@ class UserConverter(IDConverter):
 
         return result
 
+class MessageConverter(IDConverter):
+    """Converts to a :class:`discord.Message`.
+
+    The lookup strategy is as follows (in order):
+
+    1. Lookup by "{channel ID}-{message ID}" (retrieved by shift-clicking on "Copy ID")
+    2. Lookup by message ID (the message **must** be in the context channel)
+    3. Lookup by message URL
+    """
+
+    async def convert(self, ctx, argument):
+        id_regex = re.compile(r'^(?:(?P<channel_id>[0-9]{15,21})-)?(?P<message_id>[0-9]{15,21})$')
+        link_regex = re.compile(
+            r'^https?://(?:(ptb|canary)\.)?discordapp\.com/channels/'
+            r'(?:([0-9]{15,21})|(@me))'
```

Comment:
There could be a decent speed increase if we use the guild ID (the first of the 3 IDs), since the bot wouldn't have to search through the guilds. Might be nice in the future.
(Sometimes the guild ID will just be the channel ID for dumb reasons, so it's probably fine if this isn't added)
See https://github.com/Rapptz/discord.py/pull/2091/files